Version 4.3.0 breaks compilation due to NuGet packages with multiple Roslyn assemblies (e.g. `System.Text.Json`)

Our Unity project has not compiled successfully on GitHub Actions since yesterday, when `NuGetForUnity.Cli` was auto-upgraded to `4.3.0`.
When freezing our `NuGetForUnity.Cli` to 4.2.0, everything is working again.
Looking at the changelog of `4.3.0` this may be due to the changes introduced via https://github.com/GlitchEnzo/NuGetForUnity/pull/702.

Unity build error:
```
error CS9057: The analyzer assembly '/home/runner/work/unity-project/Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/System.Text.Json.SourceGeneration.dll' references version '4.4.0.0' of the compiler, which is newer than the currently running version '4.3.0.0'
```

Reducted Unity build log:
```
[1277/1477  0s] Csc Library/Bee/artifacts/2400b0aE.dag/CustomUnityAssembly.dll (+2 others)
##### CommandLine
"/home/runner/Unity/Hub/Editor/6000.0.31f1/Editor/Data/NetCoreRuntime/dotnet" exec "/home/runner/Unity/Hub/Editor/6000.0.31f1/Editor/Data/DotNetSdkRoslyn/csc.dll" /nostdlib /noconfig /shared "@Library/Bee/artifacts/2400b0aE.dag/CustomUnityAssembly.rsp" "@Library/Bee/artifacts/2400b0aE.dag/CustomUnityAssembly.rsp2"
##### Contents of Library/Bee/artifacts/2400b0aE.dag/CustomUnityAssembly.rsp
-target:library
-out:"Library/Bee/artifacts/2400b0aE.dag/CustomUnityAssembly.dll"
-refout:"Library/Bee/artifacts/2400b0aE.dag/CustomUnityAssembly.ref.dll"
-define:UNITY_6000_0_31
-define:UNITY_6000_0
-define:UNITY_6000
-define:UNITY_5_3_OR_NEWER
-define:UNITY_5_4_OR_NEWER
-define:UNITY_5_5_OR_NEWER
-define:UNITY_5_6_OR_NEWER
-define:UNITY_2017_1_OR_NEWER
-define:UNITY_2017_2_OR_NEWER
-define:UNITY_2017_3_OR_NEWER
-define:UNITY_2017_4_OR_NEWER
-define:UNITY_2018_1_OR_NEWER
-define:UNITY_2018_2_OR_NEWER
-define:UNITY_2018_3_OR_NEWER
-define:UNITY_2018_4_OR_NEWER
-define:UNITY_2019_1_OR_NEWER
-define:UNITY_2019_2_OR_NEWER
-define:UNITY_2019_3_OR_NEWER
-define:UNITY_2019_4_OR_NEWER
-define:UNITY_2020_1_OR_NEWER
-define:UNITY_2020_2_OR_NEWER
-define:UNITY_2020_3_OR_NEWER
-define:UNITY_2021_1_OR_NEWER
-define:UNITY_2021_2_OR_NEWER
-define:UNITY_2021_3_OR_NEWER
-define:UNITY_2022_1_OR_NEWER
-define:UNITY_2022_2_OR_NEWER
-define:UNITY_2022_3_OR_NEWER
-define:UNITY_2023_1_OR_NEWER
-define:UNITY_2023_2_OR_NEWER
-define:UNITY_2023_3_OR_NEWER
-define:UNITY_6000_0_OR_NEWER
-define:PLATFORM_ARCH_64
-define:UNITY_64
-define:UNITY_INCLUDE_TESTS
-define:ENABLE_AUDIO
-define:ENABLE_CACHING
-define:ENABLE_CLOTH
-define:ENABLE_MICROPHONE
-define:ENABLE_MULTIPLE_DISPLAYS
-define:ENABLE_PHYSICS
-define:ENABLE_TEXTURE_STREAMING
-define:ENABLE_VIRTUALTEXTURING
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.0/cs/ko/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.0/cs/pl/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.0/cs/pt-BR/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.0/cs/ru/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.0/cs/tr/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.0/cs/zh-Hans/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.0/cs/zh-Hant/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/System.Text.Json.SourceGeneration.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/cs/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/de/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/es/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/fr/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/it/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/ja/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/ko/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/pl/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/pt-BR/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/ru/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/tr/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/zh-Hans/System.Text.Json.SourceGeneration.resources.dll"
-analyzer:"Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/zh-Hant/System.Text.Json.SourceGeneration.resources.dll"
"Assets/_Framework/TestsSupport/Scripts/Runtime/ObservableTracker.cs"
-langversion:9.0
/deterministic
/optimize+
/debug:portable
/nologo
/RuntimeMetadataVersion:v4.0.30319
/nowarn:0169
/nowarn:0649
/nowarn:0282
/nowarn:1701
/nowarn:1702
-warnaserror+
-nowarn:0649
/utf8output
/preferreduilang:en-US
/additionalfile:"Library/Bee/artifacts/2400b0aE.dag/CustomUnityAssembly.UnityAdditionalFile.txt"
##### Custom Environment Variables
DOTNET_MULTILEVEL_LOOKUP=0
##### ExitCode
1
##### Output
error CS9057: The analyzer assembly '/home/runner/work/unity-project/Packages/nuget-packages/InstalledPackages/System.Text.Json.7.0.3/analyzers/dotnet/roslyn4.4/cs/System.Text.Json.SourceGeneration.dll' references version '4.4.0.0' of the compiler, which is newer than the currently running version '4.3.0.0'.
```

`Packages/nuget-packages/NuGet.config`:
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
  </packageSources>
  <disabledPackageSources />
  <packageSourceCredentials />
  <activePackageSource>
    <add key="All" value="(Aggregate source)" />
  </activePackageSource>
  <config>
    <add key="packageInstallLocation" value="InPackagesFolder" />
    <add key="repositoryPath" value="./Packages" />
    <add key="PackagesConfigDirectoryPath" value="." />
    <add key="slimRestore" value="true" />
    <add key="PreferNetStandardOverNetFramework" value="true" />
  </config>
</configuration>
```

Reducted `Packages/nuget-packages/NuGet.config`:
```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Segment.Analytics.CSharp" version="2.5.0" manuallyInstalled="true" />
  <package id="Serialization.NET" version="1.4.1" />
  <package id="System.Text.Json" version="7.0.3" />
</packages>
```

--- a/src/NuGetForUnity.Cli/Program.cs
+++ b/src/NuGetForUnity.Cli/Program.cs
@@ -98,7 +98,12 @@ namespace NuGetForUnity.Cli
                         continue;
                     }
 
-                    if (!AnalyzerHelper.ShouldEnableRoslynAnalyzer(analyzerDllPath))
+                    // In Unity 2021 the errors that RoslynAnalyzer DLL's can't be imported because it has multiple files with the same name are blocking
+                    // even before our AssetPostprocessor can change the import settings so Unity knows that the DLL is a RoslynAnalyzer.
+                    // To bypass this error we generate the .dll.meta files with the RoslynAnalyzer label and the disable for platform configuration.
+                    // A alternative could be to delete the duplicate .resources.dll files when restoring,
+                    // but this would require to decide which user language to keep.
+                    if (int.Parse(Application.unityVersion.Split('.').First()) > 2021 && !AnalyzerHelper.ShouldEnableRoslynAnalyzer(analyzerDllPath))
                     {
                         continue;
                     }

--- a/src/NuGetForUnity.Cli/Program.cs
+++ b/src/NuGetForUnity.Cli/Program.cs
@@ -7,6 +7,7 @@ using System.Reflection;
 using System.Text;
 using NugetForUnity;
 using NugetForUnity.Configuration;
+using NugetForUnity.Helper;
 using UnityEngine;
 
 namespace NuGetForUnity.Cli
@@ -93,6 +94,11 @@ namespace NuGetForUnity.Cli
                 {
                     var analyzerDllMetaPath = $"{analyzerDllPath}.meta";
                     if (File.Exists(analyzerDllMetaPath))
+                    {
+                        continue;
+                    }
+
+                    if (!AnalyzerHelper.ShouldEnableRoslynAnalyzer(analyzerDllPath))
                     {
                         continue;
                     }

--- a/src/NuGetForUnity.Cli/Program.cs
+++ b/src/NuGetForUnity.Cli/Program.cs
@@ -98,15 +98,7 @@ namespace NuGetForUnity.Cli
                         continue;
                     }
 
-                    // In Unity 2021 the errors that RoslynAnalyzer DLL's can't be imported because it has multiple files with the same name are blocking
-                    // even before our AssetPostprocessor can change the import settings so Unity knows that the DLL is a RoslynAnalyzer.
-                    // To bypass this error we generate the .dll.meta files with the RoslynAnalyzer label and the disable for platform configuration.
-                    // A alternative could be to delete the duplicate .resources.dll files when restoring,
-                    // but this would require to decide which user language to keep.
-                    if (int.Parse(Application.unityVersion.Split('.').First()) > 2021 && !AnalyzerHelper.ShouldEnableRoslynAnalyzer(analyzerDllPath))
-                    {
-                        continue;
-                    }
+                    var isSupportedRoslynAnalyzer = AnalyzerHelper.ShouldEnableRoslynAnalyzer(analyzerDllPath);
 
                     utf8NoBom ??= new UTF8Encoding(false);
                     File.WriteAllText(

--- a/src/NuGetForUnity.Cli/Program.cs
+++ b/src/NuGetForUnity.Cli/Program.cs
@@ -101,17 +101,17 @@ namespace NuGetForUnity.Cli
                     var isSupportedRoslynAnalyzer = AnalyzerHelper.ShouldEnableRoslynAnalyzer(analyzerDllPath);
                     var labelsForAsset = isSupportedRoslynAnalyzer ?
                         """
-
+                        labels:
                         - RoslynAnalyzer
                         """ :
-                        "[]";
+                        "labels: []";
                     utf8NoBom ??= new UTF8Encoding(false);
                     File.WriteAllText(
                         analyzerDllMetaPath,
                         $"""
                          fileFormatVersion: 2
                          guid: {Guid.NewGuid():N}
-                         labels: {isSupportedRoslynAnalyzer}
+                         {labelsForAsset}
                          PluginImporter:
                            platformData:
                            - first:

--- a/src/NuGetForUnity.Cli/Program.cs
+++ b/src/NuGetForUnity.Cli/Program.cs
@@ -99,15 +99,19 @@ namespace NuGetForUnity.Cli
                     }
 
                     var isSupportedRoslynAnalyzer = AnalyzerHelper.ShouldEnableRoslynAnalyzer(analyzerDllPath);
+                    var labelsForAsset = isSupportedRoslynAnalyzer ?
+                        """
 
+                        - RoslynAnalyzer
+                        """ :
+                        "[]";
                     utf8NoBom ??= new UTF8Encoding(false);
                     File.WriteAllText(
                         analyzerDllMetaPath,
                         $"""
                          fileFormatVersion: 2
                          guid: {Guid.NewGuid():N}
-                         labels:
-                         - RoslynAnalyzer
+                         labels: {isSupportedRoslynAnalyzer}
                          PluginImporter:
                            platformData:
                            - first:

--- a/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
+++ b/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
@@ -62,7 +62,7 @@ namespace NugetForUnity.Helper
                     var maxMatchingVersion = allEnabledRoslynVersions.Max();
                     if (!allEnabledRoslynVersions.Contains(assetRoslynVersion) || assetRoslynVersion < maxMatchingVersion)
                     {
-                        enableRoslynAnalyzer = false;
+                        return false;
                     }
                 }
             }

--- a/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
+++ b/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
@@ -33,7 +33,6 @@ namespace NugetForUnity.Helper
         /// <returns>True if the label should be added, false otherwise.</returns>
         public static bool ShouldEnableRoslynAnalyzer(string path)
         {
-            var enableRoslynAnalyzer = true;
 
             // The nuget package can contain analyzers for multiple Roslyn versions.
             // In that case, for the same package, the most recent version must be chosen out of those available for the current Unity version.

--- a/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
+++ b/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
@@ -1,0 +1,111 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using JetBrains.Annotations;
+using NugetForUnity.Models;
+
+namespace NugetForUnity.Helper
+{
+    /// <summary>
+    ///     Helper class for analyzers.
+    /// </summary>
+    internal static class AnalyzerHelper
+    {
+        /// <summary>
+        ///     Folder used to store Roslyn-Analyzers inside NuGet packages.
+        /// </summary>
+        private const string AnalyzersFolderName = "analyzers";
+
+        /// <summary>
+        ///     Name of the root folder containing dotnet analyzers.
+        /// </summary>
+        private static readonly string AnalyzersRoslynVersionsFolderName = Path.Combine(AnalyzersFolderName, "dotnet");
+
+        /// <summary>
+        ///     Prefix for the path of dll's of roslyn analyzers.
+        /// </summary>
+        private static readonly string AnalyzersRoslynVersionSubFolderPrefix = Path.Combine(AnalyzersRoslynVersionsFolderName, "roslyn");
+
+        /// <summary>
+        ///     Determine if "RoslynAnalyzer" label should be added to an analyzer
+        /// </summary>
+        /// <param name="path">The path to the analyzer</param>
+        /// <returns>True if the label should be added, false otherwise.</returns>
+        public static bool ShouldEnableRoslynAnalyzer(string path)
+        {
+            var enableRoslynAnalyzer = true;
+
+            // The nuget package can contain analyzers for multiple Roslyn versions.
+            // In that case, for the same package, the most recent version must be chosen out of those available for the current Unity version.
+            var assetPath = Path.GetFullPath(path);
+            var assetRoslynVersion = GetRoslynVersionNumberFromAnalyzerPath(assetPath);
+            if (assetRoslynVersion != null)
+            {
+                var maxSupportedRoslynVersion = GetMaxSupportedRoslynVersion();
+                if (maxSupportedRoslynVersion == null)
+                {
+                    // the current unity version doesn't support roslyn analyzers
+                    enableRoslynAnalyzer = false;
+                }
+                else
+                {
+                    var versionPrefixIndex = assetPath.IndexOf(AnalyzersRoslynVersionsFolderName, StringComparison.Ordinal);
+                    var analyzerVersionsRootDirectoryPath = Path.Combine(
+                        assetPath.Substring(0, versionPrefixIndex),
+                        AnalyzersRoslynVersionsFolderName);
+                    var analyzersFolders = Directory.EnumerateDirectories(analyzerVersionsRootDirectoryPath);
+                    var allEnabledRoslynVersions = analyzersFolders.Select(GetRoslynVersionNumberFromAnalyzerPath)
+                        .Where(version => version != null && version.CompareTo(maxSupportedRoslynVersion) <= 0)
+                        .ToArray();
+
+                    // If most recent valid analyzers exist elsewhere, don't add label `RoslynAnalyzer`
+                    var maxMatchingVersion = allEnabledRoslynVersions.Max();
+                    if (!allEnabledRoslynVersions.Contains(assetRoslynVersion) || assetRoslynVersion < maxMatchingVersion)
+                    {
+                        enableRoslynAnalyzer = false;
+                    }
+                }
+            }
+
+            return enableRoslynAnalyzer;
+        }
+
+        [CanBeNull]
+        private static NugetPackageVersion GetRoslynVersionNumberFromAnalyzerPath(string analyzerAssetPath)
+        {
+            var versionPrefixStartIndex = analyzerAssetPath.IndexOf(AnalyzersRoslynVersionSubFolderPrefix, StringComparison.Ordinal);
+            if (versionPrefixStartIndex < 0)
+            {
+                return null;
+            }
+
+            var versionStartIndex = versionPrefixStartIndex + AnalyzersRoslynVersionSubFolderPrefix.Length;
+            var separatorIndex = analyzerAssetPath.IndexOf(Path.DirectorySeparatorChar, versionStartIndex);
+            var versionLength = separatorIndex >= 0 ? separatorIndex - versionStartIndex : analyzerAssetPath.Length - versionStartIndex;
+            var versionString = analyzerAssetPath.Substring(versionStartIndex, versionLength);
+            return string.IsNullOrEmpty(versionString) ? null : new NugetPackageVersion(versionString);
+        }
+
+        [CanBeNull]
+        private static NugetPackageVersion GetMaxSupportedRoslynVersion()
+        {
+            var unityVersion = UnityVersion.Current;
+            if (unityVersion >= new UnityVersion(2022, 3, 12, 'f', 1))
+            {
+                return new NugetPackageVersion("4.3.0");
+            }
+
+            if (unityVersion >= new UnityVersion(2022, 2, 1, 'f', 1))
+            {
+                return new NugetPackageVersion("4.1.0");
+            }
+
+            if (unityVersion >= new UnityVersion(2021, 2, 1, 'f', 1))
+            {
+                return new NugetPackageVersion("3.8.0");
+            }
+
+            return null;
+        }
+    }
+}

--- a/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
+++ b/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
@@ -67,7 +67,7 @@ namespace NugetForUnity.Helper
                 }
             }
 
-            return enableRoslynAnalyzer;
+            return true;
         }
 
         [CanBeNull]

--- a/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
+++ b/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
@@ -45,7 +45,7 @@ namespace NugetForUnity.Helper
                 if (maxSupportedRoslynVersion == null)
                 {
                     // the current unity version doesn't support roslyn analyzers
-                    enableRoslynAnalyzer = false;
+                    return false;
                 }
                 else
                 {

--- a/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
+++ b/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs
@@ -33,7 +33,6 @@ namespace NugetForUnity.Helper
         /// <returns>True if the label should be added, false otherwise.</returns>
         public static bool ShouldEnableRoslynAnalyzer(string path)
         {
-
             // The nuget package can contain analyzers for multiple Roslyn versions.
             // In that case, for the same package, the most recent version must be chosen out of those available for the current Unity version.
             var assetPath = Path.GetFullPath(path);
@@ -46,23 +45,19 @@ namespace NugetForUnity.Helper
                     // the current unity version doesn't support roslyn analyzers
                     return false;
                 }
-                else
-                {
-                    var versionPrefixIndex = assetPath.IndexOf(AnalyzersRoslynVersionsFolderName, StringComparison.Ordinal);
-                    var analyzerVersionsRootDirectoryPath = Path.Combine(
-                        assetPath.Substring(0, versionPrefixIndex),
-                        AnalyzersRoslynVersionsFolderName);
-                    var analyzersFolders = Directory.EnumerateDirectories(analyzerVersionsRootDirectoryPath);
-                    var allEnabledRoslynVersions = analyzersFolders.Select(GetRoslynVersionNumberFromAnalyzerPath)
-                        .Where(version => version != null && version.CompareTo(maxSupportedRoslynVersion) <= 0)
-                        .ToArray();
 
-                    // If most recent valid analyzers exist elsewhere, don't add label `RoslynAnalyzer`
-                    var maxMatchingVersion = allEnabledRoslynVersions.Max();
-                    if (!allEnabledRoslynVersions.Contains(assetRoslynVersion) || assetRoslynVersion < maxMatchingVersion)
-                    {
-                        return false;
-                    }
+                var versionPrefixIndex = assetPath.IndexOf(AnalyzersRoslynVersionsFolderName, StringComparison.Ordinal);
+                var analyzerVersionsRootDirectoryPath = Path.Combine(assetPath.Substring(0, versionPrefixIndex), AnalyzersRoslynVersionsFolderName);
+                var analyzersFolders = Directory.EnumerateDirectories(analyzerVersionsRootDirectoryPath);
+                var allEnabledRoslynVersions = analyzersFolders.Select(GetRoslynVersionNumberFromAnalyzerPath)
+                    .Where(version => version != null && version.CompareTo(maxSupportedRoslynVersion) <= 0)
+                    .ToArray();
+
+                // If most recent valid analyzers exist elsewhere, don't add label `RoslynAnalyzer`
+                var maxMatchingVersion = allEnabledRoslynVersions.Max();
+                if (!allEnabledRoslynVersions.Contains(assetRoslynVersion) || assetRoslynVersion < maxMatchingVersion)
+                {
+                    return false;
                 }
             }
 

--- a/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs.meta
+++ b/src/NuGetForUnity/Editor/Helper/AnalyzerHelper.cs.meta
@@ -1,0 +1,11 @@
+fileFormatVersion: 2
+guid: 7abfc0bbece578541aa4b670879d6dc2
+MonoImporter:
+  externalObjects: {}
+  serializedVersion: 2
+  defaultReferences: []
+  executionOrder: 0
+  icon: {instanceID: 0}
+  userData: 
+  assetBundleName: 
+  assetBundleVariant: 

--- a/src/NuGetForUnity/Editor/NugetAssetPostprocessor.cs
+++ b/src/NuGetForUnity/Editor/NugetAssetPostprocessor.cs
@@ -51,16 +51,6 @@ namespace NugetForUnity
         /// </summary>
         private const string RoslynAnalyzerLabel = "RoslynAnalyzer";
 
-        /// <summary>
-        ///     Name of the root folder containing dotnet analyzers.
-        /// </summary>
-        private static readonly string AnalyzersRoslynVersionsFolderName = Path.Combine(AnalyzersFolderName, "dotnet");
-
-        /// <summary>
-        ///     Prefix for the path of dll's of roslyn analyzers.
-        /// </summary>
-        private static readonly string AnalyzersRoslynVersionSubFolderPrefix = Path.Combine(AnalyzersRoslynVersionsFolderName, "roslyn");
-
         private static readonly List<BuildTarget> NonObsoleteBuildTargets = typeof(BuildTarget).GetFields(BindingFlags.Public | BindingFlags.Static)
             .Where(fieldInfo => fieldInfo.GetCustomAttribute(typeof(ObsoleteAttribute)) == null)
             .Select(fieldInfo => (BuildTarget)fieldInfo.GetValue(null))
@@ -280,44 +270,6 @@ namespace NugetForUnity
             return ConfigurationManager.NugetConfigFile.RepositoryPath + Path.DirectorySeparatorChar;
         }
 
-        [CanBeNull]
-        private static NugetPackageVersion GetMaxSupportedRoslynVersion()
-        {
-            var unityVersion = UnityVersion.Current;
-            if (unityVersion >= new UnityVersion(2022, 3, 12, 'f', 1))
-            {
-                return new NugetPackageVersion("4.3.0");
-            }
-
-            if (unityVersion >= new UnityVersion(2022, 2, 1, 'f', 1))
-            {
-                return new NugetPackageVersion("4.1.0");
-            }
-
-            if (unityVersion >= new UnityVersion(2021, 2, 1, 'f', 1))
-            {
-                return new NugetPackageVersion("3.8.0");
-            }
-
-            return null;
-        }
-
-        [CanBeNull]
-        private static NugetPackageVersion GetRoslynVersionNumberFromAnalyzerPath(string analyzerAssetPath)
-        {
-            var versionPrefixStartIndex = analyzerAssetPath.IndexOf(AnalyzersRoslynVersionSubFolderPrefix, StringComparison.Ordinal);
-            if (versionPrefixStartIndex < 0)
-            {
-                return null;
-            }
-
-            var versionStartIndex = versionPrefixStartIndex + AnalyzersRoslynVersionSubFolderPrefix.Length;
-            var separatorIndex = analyzerAssetPath.IndexOf(Path.DirectorySeparatorChar, versionStartIndex);
-            var versionLength = separatorIndex >= 0 ? separatorIndex - versionStartIndex : analyzerAssetPath.Length - versionStartIndex;
-            var versionString = analyzerAssetPath.Substring(versionStartIndex, versionLength);
-            return string.IsNullOrEmpty(versionString) ? null : new NugetPackageVersion(versionString);
-        }
-
         private static string[] ModifyImportSettingsOfRoslynAnalyzer([NotNull] PluginImporter plugin)
         {
             plugin.SetCompatibleWithAnyPlatform(false);
@@ -327,42 +279,16 @@ namespace NugetForUnity
                 plugin.SetExcludeFromAnyPlatform(platform, false);
             }
 
-            var enableRoslynAnalyzer = true;
-
-            // The nuget package can contain analyzers for multiple Roslyn versions.
-            // In that case, for the same package, the most recent version must be chosen out of those available for the current Unity version.
-            var assetPath = Path.GetFullPath(plugin.assetPath);
-            var assetRoslynVersion = GetRoslynVersionNumberFromAnalyzerPath(assetPath);
-            if (assetRoslynVersion != null)
+            bool enableRoslynAnalyzer = AnalyzerHelper.ShouldEnableRoslynAnalyzer(plugin.assetPath);
+            if (enableRoslynAnalyzer)
             {
-                var maxSupportedRoslynVersion = GetMaxSupportedRoslynVersion();
-                if (maxSupportedRoslynVersion == null)
-                {
-                    // the current unity version doesn't support roslyn analyzers
-                    enableRoslynAnalyzer = false;
-                }
-                else
-                {
-                    var versionPrefixIndex = assetPath.IndexOf(AnalyzersRoslynVersionsFolderName, StringComparison.Ordinal);
-                    var analyzerVersionsRootDirectoryPath = Path.Combine(
-                        assetPath.Substring(0, versionPrefixIndex),
-                        AnalyzersRoslynVersionsFolderName);
-                    var analyzersFolders = Directory.EnumerateDirectories(analyzerVersionsRootDirectoryPath);
-                    var allEnabledRoslynVersions = analyzersFolders.Select(GetRoslynVersionNumberFromAnalyzerPath)
-                        .Where(version => version != null && version.CompareTo(maxSupportedRoslynVersion) <= 0)
-                        .ToArray();
-
-                    // If most recent valid analyzers exist elsewhere, don't add label `RoslynAnalyzer`
-                    var maxMatchingVersion = allEnabledRoslynVersions.Max();
-                    if (!allEnabledRoslynVersions.Contains(assetRoslynVersion) || assetRoslynVersion < maxMatchingVersion)
-                    {
-                        enableRoslynAnalyzer = false;
-                    }
-                }
+                NugetLogger.LogVerbose("Configured asset '{0}' as a Roslyn-Analyzer.", plugin.assetPath);
+                return new[] { RoslynAnalyzerLabel, ProcessedLabel };
             }
-
-            NugetLogger.LogVerbose("Configured asset '{0}' as a Roslyn-Analyzer.", plugin.assetPath);
-            return enableRoslynAnalyzer ? new[] { RoslynAnalyzerLabel, ProcessedLabel } : new[] { ProcessedLabel };
+            else
+            {
+                return new[] { ProcessedLabel };
+            }
         }
 
         private static string[] ModifyImportSettingsOfGeneralPlugin([NotNull] PackageConfig packageConfig, [NotNull] PluginImporter plugin)

--- a/src/NuGetForUnity/Editor/NugetAssetPostprocessor.cs
+++ b/src/NuGetForUnity/Editor/NugetAssetPostprocessor.cs
@@ -8,7 +8,6 @@ using System.Reflection;
 using JetBrains.Annotations;
 using NugetForUnity.Configuration;
 using NugetForUnity.Helper;
-using NugetForUnity.Models;
 using UnityEditor;
 using UnityEngine;
 using Object = UnityEngine.Object;
@@ -279,16 +278,14 @@ namespace NugetForUnity
                 plugin.SetExcludeFromAnyPlatform(platform, false);
             }
 
-            bool enableRoslynAnalyzer = AnalyzerHelper.ShouldEnableRoslynAnalyzer(plugin.assetPath);
+            var enableRoslynAnalyzer = AnalyzerHelper.ShouldEnableRoslynAnalyzer(plugin.assetPath);
             if (enableRoslynAnalyzer)
             {
                 NugetLogger.LogVerbose("Configured asset '{0}' as a Roslyn-Analyzer.", plugin.assetPath);
                 return new[] { RoslynAnalyzerLabel, ProcessedLabel };
             }
-            else
-            {
-                return new[] { ProcessedLabel };
-            }
+
+            return new[] { ProcessedLabel };
         }
 
         private static string[] ModifyImportSettingsOfGeneralPlugin([NotNull] PackageConfig packageConfig, [NotNull] PluginImporter plugin)

